x-pack/plugins/observability_solution/asset_manager/server/lib/entities/install_entity_definition.ts

```diff
@@ -0,0 +1,157 @@
+import { startTransform } from './start_transform';
+import {
+  stopAndDeleteHistoryTransform,
+  stopAndDeleteLatestTransform,
+} from './stop_and_delete_transform';
+import { uninstallEntityDefinition } from './uninstall_entity_definition';
+
+export interface InstallDefinitionParams {
+  esClient: ElasticsearchClient;
+  soClient: SavedObjectsClientContract;
+  definition: EntityDefinition;
+  logger: Logger;
+  spaceId: string;
+}
+
+export async function installEntityDefinition({
+  esClient,
+  soClient,
+  definition,
+  logger,
+  spaceId,
+}: InstallDefinitionParams): Promise<EntityDefinition> {
+  const installState = {
+    ingestPipelines: {
+      history: false,
+      latest: false,
+    },
+    transforms: {
+      history: false,
+      latest: false,
+    },
+    definition: false,
+  };
+
+  try {
+    logger.debug(`Installing definition ${JSON.stringify(definition)}`);
+    const entityDefinition = await saveEntityDefinition(soClient, definition);
+    installState.definition = true;
+
+    // install ingest pipelines
+    logger.debug(`Installing ingest pipelines for definition ${definition.id}`);
+    await createAndInstallHistoryIngestPipeline(esClient, entityDefinition, logger, spaceId);
+    installState.ingestPipelines.history = true;
+    await createAndInstallLatestIngestPipeline(esClient, entityDefinition, logger, spaceId);
+    installState.ingestPipelines.latest = true;
+
+    // install transforms
+    logger.debug(`Installing transforms for definition ${definition.id}`);
+    await createAndInstallHistoryTransform(esClient, entityDefinition, logger);
+    installState.transforms.history = true;
+    await createAndInstallLatestTransform(esClient, entityDefinition, logger);
+    installState.transforms.latest = true;
+
+    return entityDefinition;
+  } catch (e) {
+    logger.error(`Failed to install entity definition ${definition.id}`, e);
+    // Clean up anything that was successful.
+    if (installState.definition) {
+      await deleteEntityDefinition(soClient, definition, logger);
+    }
+
+    if (installState.ingestPipelines.history) {
+      await deleteHistoryIngestPipeline(esClient, definition, logger);
+    }
+    if (installState.ingestPipelines.latest) {
+      await deleteLatestIngestPipeline(esClient, definition, logger);
+    }
+
+    if (installState.transforms.history) {
+      await stopAndDeleteHistoryTransform(esClient, definition, logger);
+    }
+
+    if (installState.transforms.latest) {
+      await stopAndDeleteLatestTransform(esClient, definition, logger);
+    }
+
+    throw e;
+  }
+}
+
+export async function installBuiltInEntityDefinitions({
+  esClient,
+  soClient,
+  logger,
+  builtInDefinitions,
+  spaceId,
+}: Omit<InstallDefinitionParams, 'definition'> & { builtInDefinitions: EntityDefinition[] }) {
+  if (builtInDefinitions.length === 0) return;
+
+  logger.debug(`Starting installation of ${builtInDefinitions.length} built-in definitions`);
```

Comment:
this is kinda nitty, but we log that we are installing N definitions, then later we check to see which definitions are installed and then install the ones that aren't. so we actually maybe install fewer than N definitions. 

i wonder if it would make more sense flow-wise if we first check which definitions need installing, returning a list of uninstalled definition IDs, then we install them in a seperate step.

not a blocker, fine as it is.